heavyai/heavyai-charting
# [FE-17705] Dynamically update categorical legend domain

### :speech_balloon: Description

This adds dynamic filtering to the legend for charts using a categorical legend.  Previously the original domain and range were being passed around in this part of the code, and no attempt at retrieving an updated domain was done for legends with categorical data.  Thus the legend displayed the initial domain, regardless if bbox/crossfilters/chart filters/etc were applied.  Now, when we detect categorical data, we make a call to crossfilter for the categorical field, get the updated topN values, pull the original range value (color) applied to each of the new topN values, and provide the legend with an updated domain and range.  This fixed the issue for pointmap, but did not for choropleth, due to a crossfilter bug (feature?).  That is fixed in this accompanying immerse PR: https://github.com/heavyai/immerse/pull/9240

### :page_facing_up: Jira Issue

Closes [FE-17705](https://heavyai.atlassian.net/browse/FE-17705)

### :mag: Verification Steps

1. Create a pointmap with a categorical column as the color measure
2. Zoom around the map, and make sure the legend updates with the values per the current bbox
3. Apply crossfilter, chart filters, dashboard filters, etc, making sure that the legend updates per the applied filters

### :camera_flash: Screenshot
Dynamic legend updating with standard range (no extra colors assigned by user):

https://github.com/heavyai/heavyai-charting/assets/15268289/a7fabf4f-632f-436a-9088-c9a7b07f6aa8

Now with custom range (extra color assignments added by user):

https://github.com/heavyai/heavyai-charting/assets/15268289/a526ba4c-47ae-4356-81ed-12cb531a4417


[FE-17705]: https://heavyai.atlassian.net/browse/FE-17705?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
